src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptAsScriptingOperation.cs

```diff
@@ -0,0 +1,322 @@
+        {
+            StringBuilder sb = new StringBuilder();
+
+            foreach (var item in stringCollection)
+            {
+                sb.Append(item);
+                if (options != null && !options.NoCommandTerminator)
+                {
+                    //Ensure the batch separator is always on a new line (to avoid syntax errors)
+                    //but don't write an extra if we already have one as this can affect definitions
+                    //of objects such as Stored Procedures (see TFS#9125366)
+                    sb.AppendFormat(CultureInfo.InvariantCulture, "{0}{1}{2}",
+                        item.EndsWith(Environment.NewLine) ? string.Empty : Environment.NewLine,
+                        this.BatchTerminator,
+                        Environment.NewLine);
+                }
+                else
+                {
+                    sb.AppendFormat(CultureInfo.InvariantCulture, Environment.NewLine);
+                }
+            }
+
+            return sb.ToString();
+        }
+
+        private UrnCollection CreateUrns(ServerConnection serverConnection)
+        {
+            IEnumerable<ScriptingObject> selectedObjects = new List<ScriptingObject>(this.Parameters.ScriptingObjects);
+
+            string server = serverConnection.TrueName;
+            string database = new SqlConnectionStringBuilder(this.Parameters.ConnectionString).InitialCatalog;
+            UrnCollection urnCollection = new UrnCollection();
+            foreach (var scriptingObject in selectedObjects)
+            {
+                urnCollection.Add(scriptingObject.ToUrn(server, database));
+            }
+            return urnCollection;
+        }
+
+        private void SetScriptBehavior(ScriptingOptions options)
+        {
+            // TODO: have to add Scripting behavior to Smo ScriptingOptions class 
+            // so it would support ScriptDropAndScreate
+            switch (this.Parameters.ScriptOptions.ScriptCreateDrop)
+            {
+                case "ScriptCreate":
+                    options.ScriptDrops = false;
+                    break;
+                case "ScriptDrop":
+                    options.ScriptDrops = true;
+                    break;
+                default:
+                    options.ScriptDrops = false;
+                    break;
+
+            }
+        }
+
+        private void SetScriptingOptions(ScriptingOptions scriptingOptions)
+        {
+            scriptingOptions.AllowSystemObjects = true;
+
+            // setting this forces SMO to correctly script objects that have been renamed
+            scriptingOptions.EnforceScriptingOptions = true;
+
+            //We always want role memberships for users and database roles to be scripted
+            scriptingOptions.IncludeDatabaseRoleMemberships = true;
+
+            if (this.Parameters.ScriptOptions.ScriptCompatibilityOption == SqlScriptOptions.ScriptCompatabilityOptions.Script140Compat.ToString())
```

Comment:
Minor: this kind of large if/else can be more efficiently defined in a dictionary mapping.